File: packages/nlp-service/app/api/classify.py
"""
Classification API endpoint.
PX-878: Tiered Content Classifier
"""

import time
import logging
from typing import Optional

from fastapi import APIRouter, HTTPException, Header

from app.schemas.sensitivity import (
    ClassifyRequest,
    ClassifyResponse,
    SensitivitySegmentResult,
    SensitivityTier,
    SegmentSignals,
    SentimentSignal,
)
from app.main import get_classifier, CLASSIFICATION_COUNT


router = APIRouter()
logger = logging.getLogger(__name__)

# Confidence threshold for human review
REVIEW_CONFIDENCE_THRESHOLD = float(__import__("os").getenv("SENSITIVITY_CONFIDENCE_THRESHOLD", "0.70"))


@router.post("/classify", response_model=ClassifyResponse)
async def classify_segments(
    request: ClassifyRequest,
    x_request_id: Optional[str] = Header(None, alias="X-Request-ID"),
    x_org_id: Optional[str] = Header(None, alias="X-Org-ID"),
):
    """
    Classify transcript segments into sensitivity tiers.

    Returns classification results for each segment with confidence scores
    and signals. If any segment is classified as REDACTED or has low
    confidence, the response will indicate review is required.
    """
    start_time = time.time()

    # Validate input
    if not request.segments:
        return ClassifyResponse(
            success=True,
            segments=[],
            overall_tier=SensitivityTier.STANDARD,
            confidence=1.0,
            model_version="v1.0.0",
            requires_review=False,
            processing_time_ms=0,
        )

    # Get classifier
    classifier = get_classifier()

    # Classify each segment
    results: list[SensitivitySegmentResult] = []
    overall_tier = SensitivityTier.STANDARD
    min_confidence = 1.0
    requires_review = False
    review_reasons: list[str] = []

    for segment in request.segments:
        try:
            # Get classification
            tier, confidence, category, signals = classifier.classify(
                text=segment.text,
                speaker=segment.speaker,
            )

            # Determine if review is needed
            needs_review = False
            review_reason = None

            if confidence < REVIEW_CONFIDENCE_THRESHOLD:
                needs_review = True
                review_reason = f"Low confidence ({confidence:.0%})"
                review_reasons.append(review_reason)

            if tier == SensitivityTier.REDACTED:
                needs_review = True
                review_reason = "REDACTED content requires confirmation"
                if review_reason not in review_reasons:
                    review_reasons.append(review_reason)

            # Track metrics
            CLASSIFICATION_COUNT.labels(tier=tier.value).inc()

            # Build result
            result = SensitivitySegmentResult(
                segment_index=segment.index,
                start_time=segment.start_time,
                end_time=segment.end_time,
                text=segment.text,
                tier=tier,
                confidence=confidence,
                category=category,
                signals=signals,
                needs_review=needs_review,
                review_reason=review_reason,
            )
            results.append(result)

            # Update overall metrics
            if needs_review:
                requires_review = True

            if confidence < min_confidence:
                min_confidence = confidence

            # Determine highest risk tier
            if tier == SensitivityTier.REDACTED:
                overall_tier = SensitivityTier.REDACTED
            elif tier == SensitivityTier.RESTRICTED and overall_tier != SensitivityTier.REDACTED:
                overall_tier = SensitivityTier.RESTRICTED

        except Exception as e:
            logger.error(f"Error classifying segment {segment.index}: {e}")
            # Default to STANDARD with low confidence on error
            results.append(SensitivitySegmentResult(
                segment_index=segment.index,
                start_time=segment.start_time,
                end_time=segment.end_time,
                text=segment.text,
                tier=SensitivityTier.STANDARD,
                confidence=0.5,
                category=None,
                signals=SegmentSignals(
                    entities=[],
                    sentiment=SentimentSignal(
                        compound=0.0,
                        positive=0.0,
                        negative=0.0,
                        neutral=1.0,
                        category="NEUTRAL"
                    ),
                    taxonomy=[],
                ),
                needs_review=True,
                review_reason="Classification error - manual review required",
            ))
            requires_review = True
            min_confidence = min(min_confidence, 0.5)

    # Calculate processing time
    processing_time_ms = (time.time() - start_time) * 1000

    # Build block reason if needed
    block_reason = None
    if requires_review:
        if overall_tier == SensitivityTier.REDACTED:
            block_reason = f"REDACTED content detected with {min_confidence:.0%} confidence"
        elif min_confidence < REVIEW_CONFIDENCE_THRESHOLD:
            block_reason = f"Low confidence classification ({min_confidence:.0%})"

    logger.info(
        f"Classified {len(results)} segments: "
        f"overall_tier={overall_tier.value}, "
        f"confidence={min_confidence:.2f}, "
        f"requires_review={requires_review}, "
        f"duration={processing_time_ms:.0f}ms"
    )

    return ClassifyResponse(
        success=True,
        segments=results,
        overall_tier=overall_tier,
        confidence=min_confidence,
        model_version=classifier.version,
        requires_review=requires_review,
        block_reason=block_reason,
        processing_time_ms=processing_time_ms,
    )

File: apps/web/src/lib/services/sensitivity/classify.ts
/**
 * Main sensitivity classification function.
 * PX-878: Tiered Content Classifier
 *
 * This is the primary integration point for the call processing pipeline.
 */

import type { TranscriptSegment } from "@/lib/deepgram/transcribe";
import {
  getSensitivityClient,
  isSensitivityEnabled,
  SensitivityServiceError,
} from "./client";
import type {
  SensitivityResult,
  SensitivitySegmentResult,
  TranscriptSegment as SensitivityTranscriptSegment,
} from "./types";

/**
 * Classify transcript segments for sensitivity.
 *
 * This function:
 * 1. Transforms Deepgram segments to NLP service format
 * 2. Calls the Python NLP service for classification
 * 3. Returns classification results with review flags
 *
 * If the service is unavailable, returns a fallback result that doesn't
 * block the pipeline (logs warning and continues).
 *
 * @param segments - Transcript segments from Deepgram
 * @param orgId - Organization ID
 * @param callId - Optional call ID for tracking
 * @param conversationId - Optional conversation ID
 * @returns Sensitivity classification results
 */
export async function classifySensitivity(
  segments: TranscriptSegment[],
  orgId: string,
  callId?: string,
  conversationId?: string
): Promise<SensitivityResult> {
  // Check if sensitivity detection is enabled
  if (!isSensitivityEnabled()) {
    console.log("[Sensitivity] Service not enabled, skipping classification");
    return createFallbackResult(segments, "Service not enabled");
  }

  // Handle empty segments
  if (!segments || segments.length === 0) {
    return {
      success: true,
      segments: [],
      overallTier: "STANDARD",
      confidence: 1.0,
      modelVersion: "fallback",
      requiresReview: false,
      processingTimeMs: 0,
    };
  }

  try {
    const client = getSensitivityClient();

    // Transform Deepgram segments to NLP service format
    const nlpSegments: SensitivityTranscriptSegment[] = segments.map(
      (segment, index) => ({
        index,
        startTime: segment.startTime,
        endTime: segment.endTime,
        text: segment.text,
        speaker: segment.speaker,
      })
    );

    // Call NLP service
    const result = await client.classify(
      nlpSegments,
      orgId,
      callId,
      conversationId
    );

    console.log(
      `[Sensitivity] Classified ${segments.length} segments: ` +
      `tier=${result.overallTier}, confidence=${result.confidence.toFixed(2)}, ` +
      `requiresReview=${result.requiresReview}`
    );

    return result;
  } catch (error) {
    // Log error and return fallback
    if (error instanceof SensitivityServiceError) {
      console.error(
        `[Sensitivity] Service error: ${error.code} - ${error.message}`
      );
    } else {
      console.error(`[Sensitivity] Unexpected error:`, error);
    }

    // Track service unavailability
    // TODO: trackEvent(SENSITIVITY_EVENTS.SERVICE_UNAVAILABLE, { orgId, callId, error })

    // Return fallback that doesn't block pipeline
    return createFallbackResult(
      segments,
      error instanceof Error ? error.message : "Service unavailable"
    );
  }
}

/**
 * Create a fallback result when service is unavailable.
 * Marks all segments as STANDARD with a note about service unavailability.
 */
function createFallbackResult(
  segments: TranscriptSegment[],
  reason: string
): SensitivityResult {
  const fallbackSegments: SensitivitySegmentResult[] = segments.map(
    (segment, index) => ({
      segmentIndex: index,
      startTime: segment.startTime,
      endTime: segment.endTime,
      text: segment.text,
      tier: "STANDARD" as const,
      confidence: 0.5, // Low confidence indicates uncertainty
      signals: {
        entities: [],
        sentiment: {
          compound: 0,
          positive: 0,
          negative: 0,
          neutral: 1,
          category: "NEUTRAL" as const,
        },
        taxonomy: [],
      },
      needsReview: false,
      reviewReason: `Service unavailable: ${reason}`,
    })
  );

  return {
    success: false,
    segments: fallbackSegments,
    overallTier: "STANDARD",
    confidence: 0.5,
    modelVersion: "fallback",
    requiresReview: false,
    blockReason: undefined, // Don't block on service failure
    processingTimeMs: 0,
  };
}

/**
 * Check if a sensitivity result requires blocking the pipeline.
 *
 * Pipeline should be blocked when:
 * - REDACTED content is detected
 * - OR confidence is below threshold and content is potentially sensitive
 */
export function shouldBlockPipeline(result: SensitivityResult): boolean {
  // Don't block on service failure
  if (!result.success) {
    return false;
  }

  // Block if any segment needs review
  return result.requiresReview;
}

/**
 * Get segments that require human review.
 */
export function getSegmentsRequiringReview(
  result: SensitivityResult
): SensitivitySegmentResult[] {
  return result.segments.filter((segment) => segment.needsReview);
}

/**
 * Determine overall tier from segment results.
 * Returns the highest risk tier found.
 */
export function determineOverallTier(
  segments: SensitivitySegmentResult[]
): "STANDARD" | "RESTRICTED" | "REDACTED" {
  if (segments.some((s) => s.tier === "REDACTED")) {
    return "REDACTED";
  }
  if (segments.some((s) => s.tier === "RESTRICTED")) {
    return "RESTRICTED";
  }
  return "STANDARD";
}
